### PR TITLE
Fix/conditionals and mutations

### PR DESCRIPTION
**Pull request by MateoRojas**

### CODE REVIEW

File: examples/basic/src/main/java/io/littlehorse/examples/BasicExample.java [hunk 1]
```diff
@@ -64,36 +65,38 @@ public static void main(String[] args) throws IOException {
...
-                worker.getTaskDefName()
-            );
-        } else {
-            log.debug(
-                "Task {} does not exist, registering it",
-                worker.getTaskDefName()
-            );
-            worker.registerTaskDef();
-        }
+//
```
Comment (MateoRojas) [inline examples/basic/src/main/java/io/littlehorse/examples/BasicExample.java:68]: Ooops This is not supposed to be commented

File: sdk-go/examples/conditionals/ifelse/workflow.go [hunk 2]
```diff
@@ -23,10 +23,16 @@ func DonutWorkflow(wf *wflib.WorkflowThread) {
 	wf.DoIfElse(
 		wf.Condition(numDonuts, model.Comparator_LESS_THAN, 10),
 		func(t *wflib.WorkflowThread) {
```
Comment (MateoRojas) [inline sdk-go/examples/conditionals/ifelse/workflow.go:25]: These Mutates shouldn't be here either, I was just testing

File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/WorkflowThreadImpl.java [hunk 3]
```diff
@@ -347,10 +344,12 @@ public void doIfElse(WorkflowCondition condition, IfElseBody ifBody, IfElseBody
         lastNodeCondition = cond.getSpec();
         ifBody.body(this);
 
-        // Close off the tree. The bottom node from the ifBlock tree is also
-        // going to be the bottom node from the elseBlock.
-        addNopNode();
-        String joinerNodeName = lastNodeName;
+        String lastNodeFromIfBlockName = lastNodeName;
+        Node.Builder lastNodeFromIfBlock = spec.getNodesOrThrow(lastNodeName).toBuilder();
+        List<VariableMutation> variablesFromIfBlock = new ArrayList<>(variableMutations.size());
```
Comment (MateoRojas) [inline sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/WorkflowThreadImpl.java:349]: From line 349 to 352 could be extracted to a method like how I did it on the Python SDK

File: server/src/main/java/io/littlehorse/common/model/getable/core/wfrun/ThreadRunModel.java [hunk 4]
```diff
@@ -571,18 +571,19 @@ public void complete(Date time) {
...
         crn.status = LHStatus.COMPLETED;
-        try {
-            mutateVariables(output);
-        } catch (LHVarSubError exn) {
-            fail(
-                    new FailureModel("Failed mutating variables: " + exn.getMessage(), LHConstants.VAR_MUTATION_ERROR),
-                    eventTime);
-            return;
-        }
+        //        try {
```
Comment (MateoRojas) [inline server/src/main/java/io/littlehorse/common/model/getable/core/wfrun/ThreadRunModel.java:574]: These comments should be deleted

File: server/src/main/java/io/littlehorse/common/model/getable/core/wfrun/ThreadRunModel.java [hunk 5]
```diff
@@ -620,6 +621,40 @@ public void advanceFrom(NodeModel curNode) {
         }
     }
 
+    public void advanceFrom(VariableValueModel output, NodeModel curNode) {
```
Comment (MateoRojas) [inline server/src/main/java/io/littlehorse/common/model/getable/core/wfrun/ThreadRunModel.java:624]: I just copied the `advanceFrom` twice, so now there are two `advanceFrom` one that mutates and one that doesn't. I should find a better way of doing this without copying the exact same code twice

File: server/src/main/java/io/littlehorse/common/model/getable/global/wfspec/node/NodeModel.java [hunk 6]
```diff
@@ -48,7 +46,7 @@ public class NodeModel extends LHSerializable<Node> {
     public UserTaskNodeModel userTaskNode;
     private StartMultipleThreadsNodeModel startMultipleThreadsNode;
 
-    public List<VariableMutationModel> variableMutations;
+    //    public List<VariableMutationModel> variableMutations;
```
Comment (MateoRojas) [inline server/src/main/java/io/littlehorse/common/model/getable/global/wfspec/node/NodeModel.java:49]: This comment should be deleted

File: server/src/main/java/io/littlehorse/common/model/getable/global/wfspec/node/NodeModel.java [hunk 7]
```diff
@@ -63,9 +61,9 @@ public Node.Builder toProto() {
             out.addOutgoingEdges(o.toProto());
         }
```
Comment (MateoRojas) [inline server/src/main/java/io/littlehorse/common/model/getable/global/wfspec/node/NodeModel.java:63]: These comments should be deleted

File: server/src/main/java/io/littlehorse/common/model/getable/global/wfspec/node/NodeModel.java [hunk 8]
```diff
@@ -120,11 +118,11 @@ public void initFrom(Message p, ExecutionContext context) {
             outgoingEdges.add(edge);
         }
```
Comment (MateoRojas) [inline server/src/main/java/io/littlehorse/common/model/getable/global/wfspec/node/NodeModel.java:120]: These comments should be deleted

File: server/src/main/java/io/littlehorse/common/model/getable/global/wfspec/node/NodeModel.java [hunk 9]
```diff
@@ -179,7 +177,7 @@ public void initFrom(Message p, ExecutionContext context) {
 
     public NodeModel() {
         outgoingEdges = new ArrayList<>();
-        variableMutations = new ArrayList<>();
+        //        variableMutations = new ArrayList<>();
```
Comment (MateoRojas) [inline server/src/main/java/io/littlehorse/common/model/getable/global/wfspec/node/NodeModel.java:180]: This comment should be deleted

File: server/src/main/java/io/littlehorse/common/model/getable/global/wfspec/node/NodeModel.java [hunk 10]
```diff
@@ -188,22 +186,22 @@ public NodeModel() {
 
     public ThreadSpecModel threadSpecModel;
```
Comment (MateoRojas) [inline server/src/main/java/io/littlehorse/common/model/getable/global/wfspec/node/NodeModel.java:188]: These comments should be deleted

File: examples/basic/src/main/java/io/littlehorse/examples/BasicExample.java (hunk 1 above)
Comment (coltmcnealy-lh) [reply examples/basic/src/main/java/io/littlehorse/examples/BasicExample.java:68]: We're going to make `PutTaskDef` idempotent too (:

Comment (MateoRojas) [reply examples/basic/src/main/java/io/littlehorse/examples/BasicExample.java:68]: :tada:

File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/WorkflowThreadImpl.java [hunk 11]
```diff
@@ -69,11 +71,13 @@ final class WorkflowThreadImpl implements WorkflowThread {
     private EdgeCondition lastNodeCondition;
     private boolean isActive;
     private ThreadRetentionPolicy retentionPolicy;
+    private Queue<VariableMutation> variableMutations;
```
Comment (coltmcnealy-lh) [inline sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/WorkflowThreadImpl.java:74]: This reminds me. Variable Mutations should be ordered... for example:

```
WfRunVariable foo = wf.addVariable("foo", INT);
WfRunVariable bar = wf.addVariable("bar", INT);

// somehow foo == 0, bar == 5

wf.mutate(foo, ADD, 10);
wf.mutate(bar, ADD, foo);

// Should have: foo == 10, bar == 15
```

File: server/src/main/java/io/littlehorse/common/model/getable/core/wfrun/ThreadRunModel.java (hunk 5 above)
Comment (coltmcnealy-lh) [reply server/src/main/java/io/littlehorse/common/model/getable/core/wfrun/ThreadRunModel.java:624]: Are there clients of both `advanceFrom()` methods, and if so, should we migrate all to use the one that mutates?

**Review by coltmcnealy-lh — APPROVED** [review]
Looks good to me actually. Well done!

File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/WorkflowThreadImpl.java [hunk 12]
```diff
@@ -364,7 +365,15 @@ public void doIfElse(WorkflowCondition condition, IfElseBody ifBody, IfElseBody
 
         // The bottom node from the ifBlock tree is also
         // going to be the bottom node from the elseBlock.
+        Node.Builder lastNodeFromIfBlock = spec.getNodesOrThrow(lastNodeFromIfBlockName).toBuilder();
         Edge.Builder ifBlockEdge = Edge.newBuilder().setSinkNodeName(lastNodeName);
+
+        // If the treeRootNodeName is equal to the lastNodeFromIfBlockName it means that
+        // no node was created within the if block, thus the edge of the starting NOP should be created
+        // with the appropriate conditional
```
Comment (coltmcnealy-lh) [inline sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/WorkflowThreadImpl.java:373]: For simplicity, maybe we could just call `addNopNode()`? That would make it much less likely to have bugs in it, and we can optimize later.

**Pull request merged**